cdap-watchdog/src/main/java/co/cask/cdap/logging/meta/AbstractCheckpointManager.java

```diff
@@ -0,0 +1,148 @@
+        result.put(entry.getKey(), new Checkpoint<>(checkpoint.getOffset(), checkpoint.getMaxEventTime()));
+      }
+      return result;
+
+    }, IOException.class);
+
+    LOG.trace("Saved checkpoints for partitions {}", checkpoints);
+  }
+
+  @Override
+  public Map<Integer, Checkpoint<T>> getCheckpoint(final Set<Integer> partitions) throws IOException {
+    return TransactionRunners.run(transactionRunner, context -> {
+      Map<Integer, Checkpoint<T>> checkpoints = new HashMap<>();
+      StructuredTable table = context.getTable(StoreDefinition.LogCheckpointStore.LOG_CHECKPOINT_TABLE);
+      for (final int partition : partitions) {
+        List<Field<?>> keyFields = ImmutableList.<Field<?>>builder()
+          .addAll(getKeyFields(rowKeyPrefix, partition))
+          .build();
+        Optional<StructuredRow> optionalRow = table.read(keyFields);
+        StructuredRow row = optionalRow.orElse(null);
+        checkpoints.put(partition, fromRow(row));
+      }
+      return checkpoints;
+    }, IOException.class);
+  }
+
+  @Override
+  public Checkpoint<T> getCheckpoint(final int partition) throws IOException {
+    Checkpoint<T> checkpoint = TransactionRunners.run(transactionRunner, context -> {
+      StructuredTable table = context.getTable(StoreDefinition.LogCheckpointStore.LOG_CHECKPOINT_TABLE);
+      List<Field<?>> keyFields = ImmutableList.<Field<?>>builder()
+        .addAll(getKeyFields(rowKeyPrefix, partition))
+        .build();
+      Optional<StructuredRow> optionalRow = table.read(keyFields);
+      StructuredRow row = optionalRow.orElse(null);
+      return fromRow(row);
+    }, IOException.class);
+
+    LOG.trace("Read checkpoint {} for partition {}", checkpoint, partition);
+    return checkpoint;
+  }
+
+  private Checkpoint<T> fromRow(@Nullable StructuredRow row) throws IOException {
+    byte[] offset = null;
+    if (row != null) {
+      offset = row.getBytes(StoreDefinition.LogCheckpointStore.CHECKPOINT_FIELD);
+    }
+    return deserializeCheckpoint(offset);
+  }
+
+  /**
+   * Serialize the checkpoint.
+   * @param checkpoint checkpoint to be serialized
+   * @return serialized checkpoint byte array
+   * @throws IOException if error while serializing checkpoint
+   */
+  protected abstract byte[] serializeCheckpoint(Checkpoint<T> checkpoint) throws IOException;
+
+  /**
+   * Deserialize the checkpoint bytes. If checkpoint bytes are null, then return {@code -1} for all the numeric
+   * values of checkpoint such as maxEventTime.
+   * @param checkpoint checkpoint bytes to be deserialized
+   * @return deserialized checkpoint
+   * @throws IOException if error while deserializing checkpoint bytes
+   */
+  protected abstract Checkpoint<T> deserializeCheckpoint(@Nullable byte[] checkpoint) throws IOException;
+
+  private Collection<Field<?>> getKeyFields(String rowPrefix, int partition) {
+    return Arrays.asList(Fields.stringField(StoreDefinition.LogCheckpointStore.ROW_PREFIX_FIELD, rowPrefix),
```

Comment:
it's a bit more performant to create a List and add the fields directly instead of converting an array into a list.